datafusion/core/src/physical_optimizer/limit_pushdown.rs

```diff
@@ -0,0 +1,127 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! The [`LimitPushdown`] The LimitPushdown optimization rule is designed
+//! to improve the performance of query execution by pushing the LIMIT clause down
+//!  through the execution plan as far as possible, ideally directly
+//! to the [`CoalesceBatchesExec`]. to reduce target_batch_size This means that instead of processing
+//! a large amount of data and then applying the limit at the end,
+//! the system tries to limit the amount of data being processed throughout the execution of the query.
+
+use std::sync::Arc;
+
+use crate::physical_optimizer::utils::is_global_limit;
+use crate::physical_optimizer::PhysicalOptimizerRule;
+
+use crate::physical_plan::limit::GlobalLimitExec;
+use crate::physical_plan::ExecutionPlan;
+use datafusion_common::config::ConfigOptions;
+use datafusion_common::tree_node::{Transformed, TransformedResult, TreeNode};
+use datafusion_common::Result;
+
+use datafusion_physical_plan::coalesce_batches::CoalesceBatchesExec;
+use datafusion_physical_plan::coalesce_partitions::CoalescePartitionsExec;
+
+#[allow(missing_docs)]
+pub struct LimitPushdown {}
+
+impl LimitPushdown {
+    #[allow(missing_docs)]
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+impl PhysicalOptimizerRule for LimitPushdown {
+    fn optimize(
+        &self,
+        plan: Arc<dyn ExecutionPlan>,
+        _config: &ConfigOptions,
+    ) -> Result<Arc<dyn ExecutionPlan>> {
+        // if this node is not a global limit, then directly return
+        if !is_global_limit(&plan) {
```

Comment:
Can we extend the rule to pushdown GlobalLimit's which are not at the top of the plan?

Reply:
Definitely, this is just a draft to test the exist .slt test